patches/tModLoader/Terraria/ModLoader/Core/ModCompile.cs

```diff
@@ -54,20 +54,26 @@ public BuildingMod(TmodFile modFile, BuildProperties properties, string path) :
 
 	public static readonly string ModSourcePath = Path.Combine(Program.SavePathShared, "ModSources");
 
-	internal static string[] FindModSources()
+	internal static string[] FindModSources(bool refindAllMods, out LocalMod[] builtMods)
 	{
 		Directory.CreateDirectory(ModSourcePath);
-		return Directory.GetDirectories(ModSourcePath, "*", SearchOption.TopDirectoryOnly).Where(dir => {
+		var modSources = Directory.GetDirectories(ModSourcePath, "*", SearchOption.TopDirectoryOnly).Where(dir => {
 			var directory = new DirectoryInfo(dir);
 			return directory.Name[0] != '.' && directory.Name != "ModAssemblies" && directory.Name != "Mod Libraries";
-		}).ToArray();
+		}).ToHashSet();
+
+		// Also add mod sources defined by built .tmod files.
+		builtMods = refindAllMods ? ModOrganizer.FindAllMods() : ModOrganizer.AllFoundMods;
+		modSources.UnionWith(builtMods.Where(m => m.location == ModLocation.Local).Select(m => m.properties.modSource).Where(s => s.Length > 0));
+
+		return modSources.ToArray();
 	}
 
 	// Silence exception reporting in the chat unless actively modding.
 	public static bool activelyModding;
 	internal static DateTime recentlyBuiltModCheckTimeCutoff = DateTime.Now - TimeSpan.FromSeconds(60);
 
-	public static bool DeveloperMode => Debugger.IsAttached || Directory.Exists(ModSourcePath) && FindModSources().Length > 0;
+	public static bool DeveloperMode => Debugger.IsAttached || Directory.Exists(ModSourcePath) && FindModSources(refindAllMods: false, out _).Length > 0;
```

Comment:
Might be good to use a cached result for the `FindModSources`. Will the `Directory.Exists(ModSourcePath)` potentially be invalidated now?

Reply:
`ModSourcePath` is created upon resolving mod sources (in `FindModSources`), so it should be fine.

Reply:
I've updated it to skip the check but not create/search the directory in `FindModSources` if it doesn't exist.